src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp

```diff
@@ -247,28 +249,28 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
             for (const auto& in_desc : then_in_desc) {
                 const auto& in_indx = in_desc->m_input_index;
                 const auto& body_indx = in_desc->m_body_parameter_index;
-                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_partial_shape =
-                        then_body_params.at(body_indx)->get_partial_shape();
+                auto then_p = then_body_params.at(body_indx);
+                if (if_op->get_input_tensor(in_indx).get_partial_shape().compatible(then_p->get_partial_shape())) {
+                    PartialShape::merge_into(if_op->get_input_tensor(in_indx).m_partial_shape,
+                                             then_p->get_partial_shape());
                     is_changed = true;
                 }
                 if (if_op->get_input_tensor(in_indx).get_element_type().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_element_type =
-                        then_body_params.at(body_indx)->get_element_type();
+                    if_op->get_input_tensor(in_indx).m_element_type = then_p->get_element_type();
                     is_changed = true;
                 }
             }
             for (const auto& in_desc : else_in_desc) {
                 const auto& in_indx = in_desc->m_input_index;
                 const auto& body_indx = in_desc->m_body_parameter_index;
-                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_partial_shape =
-                        else_body_params.at(body_indx)->get_partial_shape();
+                auto else_p = else_body_params.at(body_indx);
+                if (if_op->get_input_tensor(in_indx).get_partial_shape().compatible(else_p->get_partial_shape())) {
+                    PartialShape::merge_into(if_op->get_input_tensor(in_indx).m_partial_shape,
+                                             else_p->get_partial_shape());
```

Comment:
I am not sure that this is correct. For example, let us have `If` operation with `then-body` param of shape [-1] and `else-body` param shape `3` but it does not mean that the corresponding `if` input should have a shape [3]. It should have [-1] still

Reply:
This code propagates shapes and types from body parameters to inputs of If operation, and those must have same shape and type for both bodies. The code for outputs isn't changed in this PR.

Reply:
Discussed offline and agreed to re-consider changes in this PR